### PR TITLE
Integrate curve v2 and balance v2

### PR DESCRIPTION
**Pull request by eric6311**
- integrate` curveV2` and `balancerV2` into `AMMWrapperWithPath` in public repo 
- integrate and adjust some setups in test cases (most of the test cases are directly copy and paste from private repo)

### CODE REVIEW

File: contracts/AMMWrapper.sol [hunk 1]
```diff
@@ -408,18 +410,28 @@ contract AMMWrapper is IAMMWrapper, ReentrancyGuard, BaseLibEIP712, SignatureVal
...
-        } else if (_swapMethod == 2) {
-            curve.exchange_underlying{ value: msg.value }(i, j, _takerAssetAmount, _makerAssetAmount);
+        if (_version == 1) {
+            ICurveFi curve = ICurveFi(_makerAddr);
+            if (_swapMethod == 1) {
+                curve.exchange{ value: msg.value }(i, j, _takerAssetAmount, _makerAssetAmount);
+            } else if (_swapMethod == 2) {
+                curve.exchange_underlying{ value: msg.value }(i, j, _takerAssetAmount, _makerAssetAmount);
+            }
+            return;
```
Comment (charlesjhongc) [inline contracts/AMMWrapper.sol:427]: I think tx should be reverted at here.

Comment (charlesjhongc) [reply contracts/AMMWrapper.sol:427]: Can add test case for unknown curve version/swapMethod handling.

**Review by NIC619 — DISMISSED** [review]

File: contracts/test/forkMainnet/AMMWrapperWithPath/TradeCurveV2.t.sol [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        order.takerAssetAddr = USDC_ADDRESS;
+        order.makerAddr = CURVE_USDT_POOL_ADDRESS;
+        sig = _signTrade(userPrivateKey, order);
+        payload = _genTradePayload(order, DEFAULT_FEE_FACTOR, sig, _encodeCurveData(2), new address[](0));
+
+        vm.expectRevert("AMMWrapper: Curve v2 no underlying");
+        userProxy.toAMM(payload);
+    }
+
+    function testCannotTradeCurveV2_With_MismatchSwapToken() public {
```
Comment (pilagod) [inline contracts/test/forkMainnet/AMMWrapperWithPath/TradeCurveV2.t.sol:54]: This test case naming rule could align with others

**Review by NIC619 — APPROVED** [review]

File: contracts/test/forkMainnet/AMMWrapperWithPath/TradeCurveV2.t.sol (hunk 2 above)
Comment (charlesjhongc) [reply contracts/test/forkMainnet/AMMWrapperWithPath/TradeCurveV2.t.sol]: Fixed in 10c3349

**Review by pilagod — APPROVED** [review]

**Pull request merged**